### PR TITLE
Add client.forms.update to publish form updates

### PR DESCRIPTION
**Pull request by lognaturel**
Based on #25. It just had a couple of changes left to make and I'm eager to get this out and published.

Noteworthy changes this makes:
- Removes the ability to do an update that only bumps the version. There's no reason to do this so it points to an application error and an exception is helpful.
- Removes `version_updater`. Maybe it's a nice affordance but we can add it if it comes up. Maybe just specifying a version as opposed to a transformation function would be preferred. Maybe nobody cares.
- Removes the version return. It's not clear what it would be used for. To make it complete we'd want to get the version from the response for the case where only the form definition is updated. It seems like complexity without clear payoff. We can add a return value if we see a need for it.
- Ignore XLSForm warnings. At that point they're not actionable. Warnings can always be safely ignored.

#### What has been done to verify that this works as intended?
Added tests of the high-level flow of what update does. I think those effectively document what is intentional about the implementation and cover most interesting logic.

#### Why is this the best possible solution? Were any other approaches considered?
I think this addresses user needs in the most minimal way leaving various options to expand functionality as needed.

#### How does this change affect users? Describe intentional changes to behavior and behavior that could have accidentally been affected by code changes. In other words, what are the regression risks?
No regression risk, all new.

#### Do we need any specific form for testing your changes? If so, please attach one.
No.

#### Does this change require updates to documentation? If so, please file an issue [here]( https://github.com/getodk/docs/issues/new) and include the link below.
Readme updated.

#### Before submitting this PR, please make sure you have:
- [x] included test cases for core behavior and edge cases in `tests`
- [x] run `nosetests` and verified all tests pass
- [x] run `python bin/pre_commit.py` to format / lint code
- [x] verified that any code or assets from external sources are properly credited in comments

### CODE REVIEW

**Comment by lognaturel** [comment]
Couple of quick high-level questions for you, @lindsay-stevens:
- Can `create` and `update` in `FormDraftService` ever return `False`? If so, how? It looks like any error state I can think of results in an exception.
- Do the new tests look like a reasonable pattern to you for testing user-facing methods? I don't usually love so much mocking but it feels like a good option here. An alternative would be to pass dependencies into `FormService` but I don't really see a big benefit. I feel like it's ok not to test the building block methods we've made private because they're so simple. The tests would be very repetitive.

File: pyodk/_endpoints/forms.py [hunk 1]
```diff
@@ -110,24 +110,23 @@ def update(
...
-        :return: The new version of the form.
         """
+        if definition is None and attachments is None:
+            raise PyODKError("Must specify a form definition and/or attachments.")
+
         # Start a new draft - with a new definition, if provided.
         fp_ids = {"form_id": form_id, "project_id": project_id}
         fd = FormDraftService(session=self.session, **self._default_kw())
-        if not fd.create(file_path=definition, **fp_ids):
+        if not fd.create(file_path=definition, ignore_warnings=True, **fp_ids):
```
Comment (lindsay-stevens) [inline pyodk/_endpoints/forms.py:129]: If never relevant then we can set the default on fd.create instead

File: pyodk/_endpoints/forms.py [hunk 2]
```diff
@@ -137,13 +136,8 @@ def update(
                 if not fda.upload(file_path=attach, **fp_ids):
                     raise PyODKError("Form update (attachment upload) failed.")
 
-        # Get a new version - using either a timestamp or the callback.
-        new_version = datetime.now().isoformat()
-        if version_updater is not None:
-            new_version = version_updater(self.get(**fp_ids).version)
+        new_version = datetime.now().isoformat() if definition is None else None
```
Comment (lindsay-stevens) [inline pyodk/_endpoints/forms.py:139]: I think it's better to write as a block for readability and in case other cases are added

File: pyodk/_endpoints/forms.py [hunk 3]
```diff
@@ -137,13 +136,8 @@ def update(
...
 
-        # Publish the draft.
+        # Publish the draft
```
Comment (lindsay-stevens) [inline pyodk/_endpoints/forms.py:141]: Punctuation is cool

File: pyodk/_utils/session.py [hunk 4]
```diff
@@ -115,7 +115,7 @@ def response_or_error(
             response.raise_for_status()
         except HTTPError as e:
             msg = (
-                f"The request to {url} failed."
+                f"The request to {self.base_url}{url} failed."
```
Comment (lindsay-stevens) [inline pyodk/_utils/session.py:118]: self.urljoin(url) will add the base_url and handle stripping slash. Better to use same function.

File: tests/endpoints/test_forms.py [hunk 5]
```diff
@@ -41,3 +45,121 @@ def test_get__ok(self):
                     form_id=fixture["response_data"][0]["xmlFormId"]
                 )
                 self.assertIsInstance(observed, Form)
+
+    def test_update__requires_definition_or_attachments(self):
```
Comment (lindsay-stevens) [inline tests/endpoints/test_forms.py:49]: I think the idiomatic way is self.assertRaises() context manager around the call that might raise, then check the exception args for the error message.

File: tests/endpoints/test_forms.py [hunk 6]
```diff
@@ -41,3 +45,121 @@ def test_get__ok(self):
...
+            try:
+                client.forms.update("foo")
+            except PyODKError:
+                pass
+            else:
+                self.fail("Exception expected")
+
+    @patch("pyodk._endpoints.forms.FormDraftAttachmentService")
+    @patch("pyodk._endpoints.forms.FormDraftService")
+    def test_update__definition_only__creates_then_publishes_draft(
```
Comment (lindsay-stevens) [inline tests/endpoints/test_forms.py:60]: I've re-written it using a helper function so that these mock tests can be more succinct.

**Comment by lindsay-stevens** [comment]
Hi @lognaturel - to address the above notes I've incorporated the changes into a new commit on #25.

> Can create and update in FormDraftService ever return False? If so, how? It looks like any error state I can think of results in an exception.

These return the response data `success` value. So it could be False in that case, depending on Central behaviour. It's not an exciting return value but no harm either.

> Do the new tests look like a reasonable pattern to you for testing user-facing methods?

They're useful to reinforce the expected behaviour. To try to reduce repetition I added a helper method to make all the patch mocks.

File: pyodk/_endpoints/forms.py (hunk 3 above)
Comment (lognaturel) [reply pyodk/_endpoints/forms.py:141]: 😂 I think I was going to change this comment in some way and didn't end up doing so? Will change back.
